### ISSUE TITLE
C2FFI Generates Invalid Macro File for GStreamer

### ISSUE THREAD

**Issue by Shinmera** [rpav/c2ffi#35]
I'm trying to use cl-autowrap (and thus c2ffi) to generate wrappers for GStreamer 1.0. Unfortunately, c2ffi crashes with a segfault in the process of this. I'm not sure if this is due to a fault on my behalf or if the gstreamer headers are simply too complex to grok.

The cl-autowrap call I'm using is:

```
(autowrap:c-include
 "/usr/include/gstreamer-0.10/gst/gst.h"
 :sysincludes '("/usr/include/gstreamer-0.10"
                "/usr/include/glib-2.0"
                "/usr/lib/glib-2.0/include"
                "/usr/include/libxml2")
 :trace-c2ffi cl:T)
```

Which executes the following two c2ffi calls, the first of which succeeds, and the second of which produces a long, long stream of errors which presumably originate from Clang.

```
c2ffi /usr/include/gstreamer-1.0/gst/gst.h -D null -M /tmp/tmpP44I4E0Y.tmp -A x86_64-pc-linux-gnu -i /usr/include/gstreamer-1.0 -i /usr/include/glib-2.0 -i /usr/lib/gstreamer-1.0/include -i /usr/lib/glib-2.0/include
c2ffi /tmp/tmpP44I4E0Y.tmp -o ~/gst.x86_64-pc-linux-gnu.spec -A x86_64-pc-linux-gnu -i /usr/include/gstreamer-1.0 -i /usr/include/glib-2.0 -i /usr/lib/gstreamer-1.0/include -i /usr/lib/glib-2.0/include
```

Since the error output is so long I've [attached it](https://github.com/rpav/c2ffi/files/321842/a.txt). I'm running the latest Quicklisp cl-autowrap, freshly cloned c2ffi and compiled it with Clang 3.7.

Any pointers as to what I could do to get this working would be much appreciated.

**Comment by rpav** [comment]
Hrm well crashing is definitely an issue; I don't see anything particularly telling near the top/bottom of the log though.  Usually crash issues happen on some odd declaration or combination thereof.  Unfortunately I'm not really back till Tuesday so I don't have facilities to debug this until then; however one common thing is missing headers (look for such errors).

Otherwise seeing what declaration it's crashing on is the most informative, usually the one just after the last actually written to the file.

**Comment by Shinmera** [comment]
Here's the temp files generated and what it managed to produce of the spec file. The names are a bit different because I did another clean run just to make sure I can reproduce the failure. 
[tmp5GEXGEG5.tmp.txt](https://github.com/rpav/c2ffi/files/321852/tmp5GEXGEG5.tmp.txt)
[tmpAR3FSGEY.tmp.txt](https://github.com/rpav/c2ffi/files/321853/tmpAR3FSGEY.tmp.txt)
[gst.x86_64-pc-linux-gnu.spec.txt](https://github.com/rpav/c2ffi/files/321854/gst.x86_64-pc-linux-gnu.spec.txt)

This isn't really urgent, so I don't mind waiting until you have the time to look into this.

**Comment by rpav** [comment]
Somehow looking at the log on my phone I missed the (obvious!) assert at the bottom.  I can't quite reproduce this, but I don't think I'm picking up all the includes properly yet.

**Comment by rpav** [comment]
All right, I believe this is fixed with 2c2bc76.  I was able to reproduce the bug and the fix appeared to generate "clean"/complete output without asserts.

Note that you still get a lot of errors output at the end simply due to how detecting `#define` macro values works, but this is normal (and may help you figure out why certain macros aren't getting included).
